### PR TITLE
Snowmtl: Adds support for multiple languages

### LINKED ISSUE #6509: Other languages for snowmtl
### Source name

Snow machine translation

### Source language

English

### Describe your suggested feature

Is it possible to add other languages? Basically, Snowmtl does not provide a feature to change the language, but this is possible when using the translate feature from Google Translate.

### Other details

Screenshot_2024-12-07-09-01-56-304_com android chrome

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] If this is an issue with the app itself, I should be opening an issue in the [app repository](https://github.com/mihonapp/mihon/issues/new/choose).
- [X] I have updated the app to version **[0.15.3](https://github.com/mihonapp/mihon/releases/latest)**.
- [X] I will fill out all of the requested information in this form.

### 

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/keiyoushi/extensions-source/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by choppeh**
Closes  #6509

- There's a problem when the user tries to read the same chapter in different langs.
- More langs may be supported in future versions

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [ ] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

**Comment by cuong-tran** [comment]
How is this supposed to close #6509 ?

**Comment by choppeh** [comment]
> How is this supposed to close #6509 ?

Adding translations to dialogs by language. I didn't integrate with Google Translate; I used Bing Translator as an alternative.

**Comment by cuong-tran** [comment]
> > How is this supposed to close #6509 ?
> 
> Adding translations to dialogs by language. I didn't integrate with Google Translate; I used Bing Translator as an alternative.

great. Then probably mention that in PR name or message instead of only "move to all"?

File: src/all/snowmtl/src/eu/kanade/tachiyomi/extension/all/snowmtl/Snowmtl.kt [hunk 1]
```diff
@@ -23,22 +27,33 @@ import uy.kohesive.injekt.injectLazy
...
-    override val lang = "en"
+    override val lang = source.lang
 
     override val supportsLatest = true
 
     private val json: Json by injectLazy()
 
+    private val translatorClient = network.cloudflareClient.newBuilder()
+        .rateLimit(1, 3, TimeUnit.SECONDS)
+        .build()
```
Comment (AwkwardPeak7) [inline src/all/snowmtl/src/eu/kanade/tachiyomi/extension/all/snowmtl/Snowmtl.kt:49]: wouldn't this slow down the translation too much? can we handle 429 gracefully?

File: src/all/snowmtl/src/eu/kanade/tachiyomi/extension/all/snowmtl/interceptors/TranslationInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+     */
+    private fun translationOptimized(dialogues: List<Dialog>): List<Dialog> {
+        val mapping = buildMap(dialogues)
+
+        val tokens = tokenizeAssociatedDialog(mapping).flatMap { token ->
+            translator.translate(source.origin, source.target, token).split(delimiter)
+        }
+
+        return replaceDialoguesWithTranslations(tokens, mapping)
+    }
```
Comment (AwkwardPeak7) [inline src/all/snowmtl/src/eu/kanade/tachiyomi/extension/all/snowmtl/interceptors/TranslationInterceptor.kt:59]: correct me if wrong but what essentially happening is that you merge multiple dialogs or tokens using delimeters as long as it is withing capacity of the translator and send translation request for each of these merged tokens?

**Review by AwkwardPeak7 — COMMENTED** [review]
impressive work honestly. I feel like people will start requesting other translation providers including ai based...

**Comment by cuong-tran** [comment]
Can check TachiyomiAT for how to do AI translation.

File: src/all/snowmtl/src/eu/kanade/tachiyomi/extension/all/snowmtl/Snowmtl.kt (hunk 1 above)
Comment (choppeh) [reply src/all/snowmtl/src/eu/kanade/tachiyomi/extension/all/snowmtl/Snowmtl.kt:49]: i added this setting to stop the translator from flagging the request as DDoS. (My public IP was blocked by DuckDuckGo when I tried to use their translator 😑)

File: src/all/snowmtl/src/eu/kanade/tachiyomi/extension/all/snowmtl/interceptors/TranslationInterceptor.kt (hunk 2 above)
Comment (choppeh) [reply src/all/snowmtl/src/eu/kanade/tachiyomi/extension/all/snowmtl/interceptors/TranslationInterceptor.kt:59]: > correct me if wrong but what essentially happening is that you merge multiple dialogs or tokens [..]

yes, for example, chinese manhuas have many dialogues per page

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**